libs/sdk-core/src/ldk/store/locking_store.rs

```diff
@@ -0,0 +1,206 @@
+
+        let locking_store = Self {
+            inner: store,
+            tl,
+            version: AtomicI64::new(version),
+        };
+        locking_store.lock().await?;
+
+        Ok(locking_store)
+    }
+
+    /// Refreshes the distributed lock to extend its duration.
+    ///
+    /// This method should be called periodically to maintain ownership of the lock.
+    ///
+    /// # Returns
+    /// * `Ok(SystemTime)` - The time until which the lock should be refreshed
+    /// * `Err(Error::Conflict)` - If the lock was lost or cannot be refreshed
+    /// * `Err(Error::Internal)` - If there was an internal error
+    pub async fn refresh_lock(&self) -> Result<SystemTime, Error> {
+        let locked_until = self.lock().await?;
+        Ok(locked_until - Duration::from_secs(Self::REFRESH_WINDOW_SECS))
+    }
+
+    async fn lock(&self) -> Result<SystemTime, Error> {
+        let lock_data = self.tl.lock().await.next_lock();
+        let value = lock_data
+            .encode()
+            .map_err(|()| Error::Internal("Failed to encode lock_data".to_string()))?;
+        self.inner
+            .put(
+                Self::KEY.to_string(),
+                value,
+                self.version.load(Ordering::SeqCst),
+            )
+            .await?;
+        self.tl.lock().await.update_lock(&lock_data);
+        self.version.fetch_add(1, Ordering::SeqCst);
+        Ok(lock_data.locked_until)
+    }
+
+    /// Releases the distributed lock.
+    ///
+    /// This method should be called when the instance is done using the store
+    /// to allow other instances to acquire the lock. If the lock has already
+    /// expired or been lost, this method will still succeed.
+    ///
+    /// # Returns
+    /// * `Ok(())` - If the lock was successfully released or was already expired
+    /// * `Err(Error::Internal)` - If there was an internal error during release
+    pub async fn unlock(&self) -> Result<(), Error> {
+        if let Some(lock_data) = self.tl.lock().await.unlock() {
+            let value = lock_data
+                .encode()
+                .map_err(|()| Error::Internal("Failed to encode lock_data".to_string()))?;
+            self.inner
+                .put(
+                    Self::KEY.to_string(),
+                    value,
+                    self.version.load(Ordering::SeqCst),
+                )
+                .await?
+        }
+        Ok(())
+    }
+
+    async fn ensure_locked(&self) -> Result<(), Error> {
+        if !self.tl.lock().await.is_locked() {
+            return Err(Error::Conflict("Remote lock was not aquired".to_string()));
+        }
+        Ok(())
+    }
+}
+
+/// Implementation of `VersionedStore` for `LockingStore`.
+///
+/// All store operations are wrapped with lock validation to ensure that
+/// only the instance holding the lock can perform operations on the underlying store.
+#[async_trait]
+impl<S: VersionedStore + Send + Sync> VersionedStore for LockingStore<S> {
```

Comment:
Curious why do we need this trait implementation? Isn't the intention that the app will use directly the LockingStore with lock/unlock/refresh_lock methods?